integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/DebugRunFacetBuilderDelegate.java

```diff
@@ -0,0 +1,187 @@
+        .orElse(null);
+  }
+
+  private SystemDebugFacet buildSystemDebugFacet() {
+    return SystemDebugFacet.builder()
+        .sparkDeployMode(getDeployMode())
+        .javaVersion(SystemUtils.JAVA_VERSION)
+        .javaVendor(SystemUtils.JAVA_VENDOR)
+        .osArch(SystemUtils.OS_ARCH)
+        .osName(SystemUtils.OS_NAME)
+        .osVersion(SystemUtils.OS_VERSION)
+        .userLanguage(SystemUtils.USER_LANGUAGE)
+        .userTimezone(SystemUtils.USER_TIMEZONE)
+        .build();
+  }
+
+  private LogicalPlanDebugFacet buildLogicalPlanDebugFacet() {
+    return olContext
+        .getQueryExecution()
+        .map(qe -> qe.optimizedPlan())
+        .map(plan -> scanLogicalPlan(plan))
+        .map(list -> new LogicalPlanDebugFacet(list))
+        .orElse(null);
+  }
+
+  private List<LogicalPlanNode> scanLogicalPlan(LogicalPlan node) {
+    List<LogicalPlanNode> result = new ArrayList<>();
+    LogicalPlanNodeBuilder builder =
+        LogicalPlanNode.builder().id(nodeId(node)).desc(node.toString());
+
+    if (node.children() != null) {
+      builder.children(
+          ScalaConversionUtils.fromSeq(node.children()).stream()
+              .map(this::nodeId)
+              .collect(Collectors.toList()));
+    }
+    result.add(builder.build());
+    if (node.children() != null) {
+      ScalaConversionUtils.fromSeq(node.children()).stream()
+          .forEach(child -> result.addAll(scanLogicalPlan(child)));
+    }
+    return result;
+  }
+
+  private String nodeId(LogicalPlan node) {
+    return String.format("%s@%s", node.nodeName(), node.hashCode());
+  }
+
+  private String getDeployMode() {
+    return Optional.ofNullable(olContext.getSparkContext()).map(sc -> sc.deployMode()).orElse(null);
+  }
+
+  private String getSparkConfOrNull(String confKey) {
+    return Optional.ofNullable(olContext.getSparkContext())
+        .map(sc -> sc.conf())
+        .map(c -> c.get(confKey, null))
+        .orElse(null);
+  }
+
+  private Map<String, String> getOpenLineageConfig() {
+    return Optional.ofNullable(olContext.getSparkContext())
+        .map(sc -> sc.conf())
+        .map(conf -> conf.getAllWithPrefix("spark.openlineage."))
+        .map(arr -> Arrays.stream(arr))
+        .map(
+            stream ->
+                stream
+                    .filter(tuple -> !tuple._1().toLowerCase().contains("key"))
+                    .collect(
+                        Collectors.<Tuple2<String, String>, String, String>toMap(
+                            t -> t._1(), t -> t._2())))
+        .orElse(Collections.emptyMap());
+  }
+
+  private String getCatalogClass() {
+    return olContext
+        .getSparkSession()
+        .map(sparkSession -> sparkSession.catalog())
+        .map(catalog -> catalog.getClass().getCanonicalName())
+        .orElse(null);
+  }
+
+  private boolean isOnClassPath(String aClass) {
+    try {
+      this.getClass().getClassLoader().loadClass(aClass);
```

Comment:
isn't `Class.forName()` more appropriate here? Does it need to be in the same ClassLoader?

Reply:
We're not consistent across the codebase and have both approaches. (https://stackoverflow.com/questions/8100376/class-forname-vs-classloader-loadclass-which-to-use-for-dynamic-loading gives more details on differences)

"getClassLoader" uses `system` classloader (which is overridable) while `Class.forName` initialises the class (runs static methods).

Reply:
I've merged the issues but I would be happy to continue discussion on that which can possibly lead to other PR or issue created.